keps/sig-api-machinery/2887-openapi-enum-types/README.md

```diff
@@ -0,0 +1,834 @@
+
+In order to convey that the `Protocol` type is an enum, every field of the type 
+have to comment its values like `ContainerPort` does. The same goes for validation - 
+protocol being one of "TCP", "UDP", and "SCTP" is duplicated among the code base.
+
+Tools and UIs can also benefit from the additional piece of information. For example,
+Protocol can be shown as a dropdown of possible values instead of a textbox.
+The user no longer needs to see an error of invalid value after submits the request.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Add an annotation that indicates a field as an enum.
+- Make the Kubernetes OpenAPI generator recognize the annotation.
+- Auto-detect possible values of an enum type.
+- Detect and annotate enum types in all built-in types. 
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+This KEP does not require removal of existing validation of enum types. While OpenAPI-based
+of enum types can simplify this type of validation, retiring custom-made validations is out
+of scope.
+
+This proposal does not intend to change the enum syntax for third-party tools like `kubebuilder`.
+The implementation should only suggest changes to internal APIs. 
+
+## Proposal
+
+As part of the OpenAPI v3 KEP, when generating OpenAPI schema,
+the de facto enum types should be reflected as enums in the resulting schema.
+We define enum types with following properties:
+- All enum types are closed, i.e., the type includes a finite number of possible values.
+- All enum types should be string. There can be no enum integers, floats, etc
+
+Additionally, This KEP makes the following assumptions on the implementation:
+- All enum types have their own type alias. i.e. , no raw string enums.
+- The canonical definition of a type is written in go and follows the standard 
+  Kubernetes convention.
+
+### Enum Pattern
+
+The Enum Pattern of a type, appearing in a go source file, is defined as follows:
+```go
+package internal
+
+// FooType define sa type alias, underlying type must be string
+type FooType string
+
+// in the same package, define constants with the alias
+const (
+    BarConst FooType = "Bar"
+    BazConst FooType = "Baz"
+)
+
+// in one or more resources or nested objects, one of the field is of 
+// the defined alias type.
+
+// ExampleResource should refer FooType
+type ExampleResource struct {
+    FooField FooType
+}
+
+```
+Enum Type is the type that follows such pattern (`FooType` in the example above).
+Enum Values are the possible values of the Enum Type (`Bar` and `Baz` are Enum Values for `FooType`).
+
+### Enum Tag
+The Enum Tag is the annotation `// +enum` that must be in the closest comment blocks
+of an enum type following Enum Pattern.
+
+```go
+package internal
+
+// FooType define sa type alias, underlying type must be string
+// +enum
```

Comment:
this auto-detection considerably diverges from kubebuilder. I see the point though that it avoid errors, and is more convenient. As written above, kubebuilder above should be pre-req for this extension.

Reply:
updated the section about kube-builder